src/swupd/files.go

```diff
@@ -0,0 +1,130 @@
+package swupd
+
+import (
+	"fmt"
+	"os"
+)
+
+// File represents an entry in a manifest
+type File struct {
+	Name        string
+	Hash        string
+	Version     uint32
+
+	// flags
+	Type        string
+	Status      string
+	Modifier    string
+	Rename      bool
+
+	// renames
+	RenameScore uint16
+	RenamePeer  *File
+
+	Info        os.FileInfo
+	DeltaPeer   *File
+}
+
+// set file type based on flag
+// F: file
+// D: directory
+// L: link
+// M: manifest
+// .: none (may be deleted/ghosted)
+func setFileTypeFromFlag(flag string, f *File) error {
+	switch flag {
+	case "F":
+		f.Type = "file"
+	case "D":
+		f.Type = "directory"
+	case "L":
+		f.Type = "link"
+	case "M":
+		f.Type = "manifest"
+	case ".":
+	default:
+		return fmt.Errorf("invalid file type flag: %v", flag)
+	}
+
+	return nil
+}
+
+// set status based on flag
+// d: deleted
+// g: ghosted
+// .: exists
+func setStatusFromFlag(flag string, f *File) error {
+	switch flag {
+	case "d":
+		f.Status = "deleted"
+	case "g":
+		f.Status = "ghosted"
+	case ".":
+	default:
+		return fmt.Errorf("invalid file status flag: %v", flag)
+	}
+
+	return nil
+}
+
+// set modifier from string
+// C: config
+// s: state
+// b: boot
+func setModifierFromFlag(flag string, f *File) error {
+	switch flag {
+	case "C":
+		f.Modifier = "config"
+	case "s":
+		f.Modifier = "state"
+	case "b":
+		f.Modifier = "boot"
+	case ".":
+	default:
+		return fmt.Errorf("invalid file modifier flag: %v", flag)
+	}
+
+	return nil
+}
+
+// set rename flag from string
+// r: this is a rename-from or renamed-to file
+func setRenameFromFlag(flag string, f *File) error {
+	switch flag {
+	case "r":
+		f.Rename = true
+	case ".":
+		f.Rename = false
+	default:
+		return fmt.Errorf("invalid file rename flag: %v", flag)
+	}
+
+	return nil
+}
+
+// set file flags from string
+func (f *File) SetFlags(flags string) error {
+	if len(flags) != 4 {
+		return fmt.Errorf("invalid number of flags: %v", flags)
+	}
+
+	var err error
+	// set file type
+	if err = setFileTypeFromFlag(string(flags[0]), f); err != nil {
```

Comment:
The fact that you have to cast this to a string is a strong hint that this is wrong.